### DISCUSSION TITLE
Comparison to Browserless

### DISCUSSION THREAD

**Discussion by sagikazarmark** [gotenberg/gotenberg#398] [Q&A]
Hey there!

Has anyone compared Gotenberg to Browserless Chrome? At least the features that are comparable.

I'm particularly interested in Gotenberg's scalability options (auto-scaling based on load) and whether there are any performance spikes when using Chrome based feature (there probably are) and how they are handled compared to Browserless.

I spent some time on getting Browserless running on Kubernetes and faced the following issues:
- metrics are in a weird, custom format, so it's hard to integrate into existing tooling
- since Chrome is quite CPU and memory intensive in some cases, scaling it turned out to be quite hard

Any insights would be much appreciated. Thanks!

**Comment by gulien** [comment]
Hello @sagikazarmark👋

I’m not familiar with Browserless so I cannot make a 1:1 comparison 😅 but I’ll try to give you some insights about how you may scale Gotenberg.

First, Gotenberg will start a dedicated instance of Chrome per request. Once a conversion is done, the instance is killed to mitigate memory leaking. But it also means that a lot of concurrent requests to a single Gotenberg instance will spawn a lot of Chrome instances. Nevertheless, vertical scaling works out of the box.

From my experience, CPU/memory usage will increase in such a scenario, but not exponentially though. Memory usage is also more important than CPU, but the more CPU is available, the quickest the conversions.

In order to scale horizontally your number of Gotenberg instances, you have two metrics:

1. Memory usage.
2. The Prometheus metric  `gotenberg_chromium_active_instances_count`.

I don’t think memory usage is a « good » metric here because it decreases somewhat slowly over time. I don’t know the reason, because even if there are no more Chrome processes, the memory does not free up immediately (the root cause is maybe Debian?).

The second metric is more interesting IMO, but it requires some fine tuning on your side to determine what’s the « best » number of active Chrome instances before scaling up/down Gotenberg. It also requires Prometheus 😅

**Reply by sagikazarmark** [reply]
@gulien thanks a lot for your insights! Much appreciated!

> Once a conversion is done, the instance is killed to mitigate memory leaking.

Yeah, that makes sense, especially to keep memory from leaking, but as far as I know, browserless keeps chrome processes around to spare the cost of spawning a new process (though it's written in NodeJS, don't know if there is any performance penalty there). Not that I think it's very important (how much that could really be? a couple miliseconds?), but do you have any numbers about the penalty of spawning new processes for every request?

> Memory usage is also more important than CPU, but the more CPU is available, the quickest the conversions.

In my experience, the problem with both memory and CPU is that they tend to grow in spikes, so it's rather hard to determine an appropriate limit. But I agree that memory is more important here as it can lead to processes being killed.

> I don’t think memory usage is a « good » metric

Generally speaking, I don't like memory as a metric for horizontal scaling decisions either, because it doesn't really represent how well or bad your application performs (you could be fine on memory, but could hit an upper limit of requests). So an active number of requests vs total number of supported requests is a better primary metric. Using CPU and memory can be secondary metrics affecting scaling.

Thanks again for the answers!

**Reply by gulien** [reply]
> Not that I think it's very important (how much that could really be? a couple miliseconds?), but do you have any numbers about the penalty of spawning new processes for every request?

I don’t have actual numbers, but the prevention of memory leaks outweighs the performance drawback. Gotenberg v6 has a single long-running Chrome instance and the memory usage just keeps growing over time. Also, a single Chrome instance cannot handle a lot of concurrent requests, so bulk conversions are in fact slower in v6 compared to v7 😄

Anyway, the biggest performance drawback is the validation of each HTTP request in a page against the allowed/denied lists 😅 it’s an important security feature though.
